File: scm-java/src/main/scheme/primitives/PrimitiveThreadNanosecond.java
package scheme.primitives;

import scheme.Pair;
import scheme.Primitive;
import scheme.SourcePos;

import java.lang.management.ManagementFactory;

public class PrimitiveThreadNanosecond extends Primitive {
    @Override
    public String name() {
        return "%thread-nanosecond";
    }

    @Override
    public String info() {
        return "Syntax: (%thread-nanosecond)\n" +
               "Library: (srfi 19)\n" +
               "Description: Internal primitive. Returns current thread CPU time as a pair (seconds . nanoseconds).\n" +
               "Example:\n" +
               "  (%thread-nanosecond) => (2 . 100000000)";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 0, 0);
        long nanoTime = ManagementFactory.getThreadMXBean().getCurrentThreadCpuTime();
        long seconds = nanoTime / 1_000_000_000L;
        long nanos = nanoTime % 1_000_000_000L;
        return new Pair(seconds, nanos);
    }
}

File: scm-csharp/scheme/primitives/PrimitiveThreadNanosecond.cs
using System.Diagnostics;

namespace scheme;

public class PrimitiveThreadNanosecond : Primitive
{
    public override string Name() => "%thread-nanosecond";

    public override string Info() =>
        "Syntax: (%thread-nanosecond)\n" +
        "Library: (scm core)\n" +
        "Description: Internal primitive. Returns current thread CPU time as a pair (seconds . nanoseconds). Falls back to process CPU time on platforms without per-thread measurement.\n" +
        "Example:\n" +
        "  (%thread-nanosecond) => (2 . 100000000)";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 0, 0);
        // .NET does not provide per-thread CPU time without P/Invoke;
        // fall back to process CPU time.
        long ticks = Process.GetCurrentProcess().TotalProcessorTime.Ticks;
        long seconds = ticks / TimeSpan.TicksPerSecond;
        long nanos = (ticks % TimeSpan.TicksPerSecond) * 100L;
        return new Pair(seconds, nanos);
    }
}
